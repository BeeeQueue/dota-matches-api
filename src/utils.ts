import ms, { StringValue } from "ms"

import { ok, RequestInitExStatus } from "@worker-tools/response-creators"

export enum MetaKey {
  MATCHES_FRESH = "MATCHES_FRESH",
  TEAMS_LAST_FETCHED = "TEAMS_LAST_FETCHED",
}

export const ms2s = (n: number) => Math.round(n / 1000)

export const nowSeconds = () => ms2s(Date.now())

export const seconds = (input: StringValue) => ms2s(ms(input))

const jsonInit: RequestInitExStatus = {
  headers: { "Content-Type": "application/json" },
}
export const json = <T = unknown>(data: T, init?: RequestInitExStatus) =>
  ok(
    JSON.stringify(data, null, 2),
    init == null
      ? jsonInit
      : {
          ...init,
          headers: {
            "Content-Type": "application/json",
            // eslint-disable-next-line unicorn/no-useless-fallback-in-spread
            ...(init.headers ?? {}),
          },
        },
  )

export const getCountry = (request: Request) => {
  if (request.cf != null && "country" in request.cf) {
    return request.cf.country
  }

  return "UNKNOWN"
}

export const EDGE_CACHE_TIMEOUT = 10_800 // seconds("3h")
const BROWSER_CACHE_TIMEOUT = 3600 // seconds("1h")

export const getTtl = (lastFetchedAt: number, ttl: number) => {
  const now = nowSeconds()
  const fetchedAtSeconds = ms2s(lastFetchedAt)
  const expiresAt = fetchedAtSeconds + ttl

  return Math.max(0, expiresAt - now)
}

export const getBrowserCacheTtl = (ttl: number) => Math.min(BROWSER_CACHE_TIMEOUT, ttl)

export const buildCacheResponse = <T = unknown>(data: T, lastFetched: number) => {
  const ttl = getTtl(lastFetched, EDGE_CACHE_TIMEOUT)

  return json(data, {
    headers: {
      "Cache-Control": `public, s-maxage=${ttl} max-age=${getBrowserCacheTtl(ttl)}`,
    },
  })
}
