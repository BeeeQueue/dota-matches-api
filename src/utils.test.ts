import { addSeconds } from "date-fns"
import { beforeAll, describe, expect, it, vi } from "vitest"

import { getTtl } from "./utils"

describe("getTtl", () => {
  beforeAll(() => {
    vi.setSystemTime(new Date("2022-02-02 12:00"))
  })

  it("should return 0", () => {
    // Fetched 10 seconds ago
    const fetchedAt = addSeconds(new Date(), -10).getTime()

    // Should expire 10 seconds after `fetchedAt`
    expect(getTtl(fetchedAt, 10)).toBe(0)
  })

  it("should return 30", () => {
    // Fetched 0 seconds ago
    const fetchedAt = Date.now()

    // Should expire 30 seconds after `fetchedAt`
    expect(getTtl(fetchedAt, 30)).toBe(30)
  })

  it("should return 0 because it expired", () => {
    // Fetched 20 seconds ago
    const fetchedAt = addSeconds(new Date(), -20).getTime()

    // Should expire 10 seconds after `fetchedAt`
    expect(getTtl(fetchedAt, 10)).toBe(0)
  })
})
