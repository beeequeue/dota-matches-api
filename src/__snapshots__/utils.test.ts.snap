// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseTeamsPage > correctly parses the body 1`] = `
[
  {
    "name": "1win Team",
    "url": "https://liquipedia.net/dota2/1win_Team",
  },
  {
    "name": "Aurora Gaming",
    "url": "https://liquipedia.net/dota2/Aurora_Gaming",
  },
  {
    "name": "AVULUS",
    "url": "https://liquipedia.net/dota2/AVULUS",
  },
  {
    "name": "Azure Ray",
    "url": "https://liquipedia.net/dota2/Azure_Ray",
  },
  {
    "name": "BetBoom Team",
    "url": "https://liquipedia.net/dota2/BetBoom_Team",
  },
  {
    "name": "BOOM Esports",
    "url": "https://liquipedia.net/dota2/BOOM_Esports",
  },
  {
    "name": "CDUB Esports",
    "url": "https://liquipedia.net/dota2/CDUB_Esports",
  },
  {
    "name": "Edge",
    "url": "https://liquipedia.net/dota2/Edge_(team)",
  },
  {
    "name": "Execration",
    "url": "https://liquipedia.net/dota2/Execration",
  },
  {
    "name": "Fart Studios",
    "url": "https://liquipedia.net/dota2/Fart_Studios",
  },
  {
    "name": "Gaimin Gladiators",
    "url": "https://liquipedia.net/dota2/Gaimin_Gladiators",
  },
  {
    "name": "HEROIC",
    "url": "https://liquipedia.net/dota2/HEROIC",
  },
  {
    "name": "Hokori",
    "url": "https://liquipedia.net/dota2/Hokori",
  },
  {
    "name": "HYDRA",
    "url": "https://liquipedia.net/dota2/HYDRA",
  },
  {
    "name": "Infamous",
    "url": "https://liquipedia.net/dota2/Infamous_Gaming",
  },
  {
    "name": "Infinity",
    "url": "https://liquipedia.net/dota2/Infinity",
  },
  {
    "name": "L1GA TEAM",
    "url": "https://liquipedia.net/dota2/L1GA_TEAM",
  },
  {
    "name": "Lava Esports",
    "url": "https://liquipedia.net/dota2/Lava_Esports",
  },
  {
    "name": "MOUZ",
    "url": "https://liquipedia.net/dota2/MOUZ",
  },
  {
    "name": "Myth Avenue Gaming",
    "url": "https://liquipedia.net/dota2/Myth_Avenue_Gaming",
  },
  {
    "name": "Natus Vincere",
    "url": "https://liquipedia.net/dota2/Natus_Vincere",
  },
  {
    "name": "NAVI Junior",
    "url": "https://liquipedia.net/dota2/NAVI_Junior",
  },
  {
    "name": "Nemiga Gaming",
    "url": "https://liquipedia.net/dota2/Nemiga_Gaming",
  },
  {
    "name": "Nigma Galaxy",
    "url": "https://liquipedia.net/dota2/Nigma_Galaxy",
  },
  {
    "name": "OG",
    "url": "https://liquipedia.net/dota2/OG",
  },
  {
    "name": "OG.LATAM",
    "url": "https://liquipedia.net/dota2/OG.LATAM",
  },
  {
    "name": "One Move",
    "url": "https://liquipedia.net/dota2/One_Move",
  },
  {
    "name": "PARIVISION",
    "url": "https://liquipedia.net/dota2/PARIVISION",
  },
  {
    "name": "Shopify Rebellion",
    "url": "https://liquipedia.net/dota2/Shopify_Rebellion",
  },
  {
    "name": "SIBE Team",
    "url": "https://liquipedia.net/dota2/SIBE_Team",
  },
  {
    "name": "Talon Esports",
    "url": "https://liquipedia.net/dota2/Talon_Esports",
  },
  {
    "name": "Team Falcons",
    "url": "https://liquipedia.net/dota2/Team_Falcons",
  },
  {
    "name": "Team Liquid",
    "url": "https://liquipedia.net/dota2/Team_Liquid",
  },
  {
    "name": "Team Secret",
    "url": "https://liquipedia.net/dota2/Team_Secret",
  },
  {
    "name": "Team Spirit",
    "url": "https://liquipedia.net/dota2/Team_Spirit",
  },
  {
    "name": "The MongolZ",
    "url": "https://liquipedia.net/dota2/The_MongolZ",
  },
  {
    "name": "TNC Predator",
    "url": "https://liquipedia.net/dota2/TNC_Predator",
  },
  {
    "name": "Tundra Esports",
    "url": "https://liquipedia.net/dota2/Tundra_Esports",
  },
  {
    "name": "Vici Gaming",
    "url": "https://liquipedia.net/dota2/Vici_Gaming",
  },
  {
    "name": "Virtus.pro",
    "url": "https://liquipedia.net/dota2/Virtus.pro",
  },
  {
    "name": "Wildcard",
    "url": "https://liquipedia.net/dota2/Wildcard",
  },
  {
    "name": "Winter Bear",
    "url": "https://liquipedia.net/dota2/Winter_Bear",
  },
  {
    "name": "Xtreme Gaming",
    "url": "https://liquipedia.net/dota2/Xtreme_Gaming",
  },
  {
    "name": "Yakult Brothers",
    "url": "https://liquipedia.net/dota2/Yakult_Brothers",
  },
  {
    "name": "Yangon Galacticos",
    "url": "https://liquipedia.net/dota2/Yangon_Galacticos",
  },
]
`;
